src/main/scala/xiangshan/cache/dcache/Uncache.scala

```diff
@@ -223,187 +183,300 @@ class UncacheImp(outer: Uncache)extends LazyModuleImp(outer)
-    entry.io.enableOutstanding := io.enableOutstanding
 
-   //  Enqueue
-    entry.io.req.valid := (i.U === enqPtr.value) && req.valid
-    entry.io.req.bits := req.bits
+  /******************************************************************
+   * Data Structure
+   ******************************************************************/
 
-    when (i.U === enqPtr.value) {
-      req_ready := entry.io.req.ready
+  val entries = Reg(Vec(UncacheBufferSize, new UncacheEntry))
+  val states = RegInit(VecInit(Seq.fill(UncacheBufferSize)(0.U.asTypeOf(new UncacheEntryState))))
+  val fence = RegInit(Bool(), false.B)
+  val s_idle :: s_refill_req :: s_refill_resp :: s_send_resp :: Nil = Enum(4)
+  val uState = RegInit(s_idle)
+  
+  def sizeMap[T <: Data](f: Int => T) = VecInit((0 until UncacheBufferSize).map(f))
+  def isStore(e: UncacheEntry): Bool = e.cmd === MemoryOpConstants.M_XWR
+  def isStore(x: UInt): Bool = x === MemoryOpConstants.M_XWR
+
+  // drain buffer
+  val empty = Wire(Bool())
+  val f1_needDrain = Wire(Bool())
+  val do_uarch_drain = RegNext(f1_needDrain)
+
+  val q0_entry = Wire(new UncacheEntry)
+  val q0_canSentIdx = Wire(UInt(INDEX_WIDTH.W))
+  val q0_canSent = Wire(Bool())
+
+
+  /******************************************************************
+   * uState for non-outstanding
+   ******************************************************************/
+
+  switch(uState){
+    is(s_idle){
+      when(req.fire){
+        uState := s_refill_req
+      }
+    }
+    is(s_refill_req){
+      when(mem_acquire.fire){
+        uState := s_refill_resp
+      }
+    }
+    is(s_refill_resp){
+      when(mem_grant.fire){
+        uState := s_send_resp
+      }
     }
+    is(s_send_resp){
+      when(resp.fire){
+        uState := s_idle
+      }
+    }
+  }
 
-    //  Acquire
-    entry.io.select := (i.U === issPtr.value) && Mux(entry.io.atomic, issPtr.value === deqPtr.value, !fence)
 
-    when (i.U === issPtr.value) {
-      need_fence := entry.io.atomic
+  /******************************************************************
+   * Enter Buffer
+   *  Version 0 (better timing)
+   *    e0 judge: alloc/merge write vec
+   *    e1 alloc
+   *
+   *  Version 1 (better performance)
+   *    solved in one cycle for achieving the original performance.
+   ******************************************************************/
+
+  /**
+    TODO lyq: how to merge
+    1. same addr
+    2. same cmd
+    3. valid
+    FIXME lyq: not merge now due to the following issues
+    1. load cann't be merged
+    2. how to merge store and response precisely
+  */
+
+  val e0_fire = req.fire
+  val e0_req = req.bits
+  /**
+    TODO lyq: block or wait or forward?
+    NOW: strict block by same address; otherwise: exhaustive consideration is needed.
+      - ld->ld wait
+      - ld->st forward
+      - st->ld forward
+      - st->st block
+  */
+  val e0_existSame = sizeMap(j => e0_req.addr === entries(j).addr && states(j).isValid()).asUInt.orR
```

Comment:
Should `mask` also be considered here in addition to checking if the `addr`s are the same?

Reply:
> Should `mask` also be considered here in addition to checking if the `addr`s are the same?

Including many other address comparison logics in this file.

Reply:
Yes, it can be considered to maximize the number of incoming entries.